File: frontend/src/lib/judgment.ts
// AuraNet™ deterministic judgment core — faithful port of the logic in
// SusMeter.dc.html. Pure functions only: safe to import from route handlers
// (server) and client components alike.
import memeTitles from '@/data/meme-titles.json';

export const TITLES: readonly string[] = memeTitles.final_verdict_titles;
export const RECS: readonly string[] = memeTitles.recommendations;

export interface Candidate {
  name: string;
  url: string;
}

export interface MatrixCell {
  text: string;
  accent: boolean;
  weight: 600 | 700;
}

export interface MatrixRow {
  label: string;
  note: string;
  cells: MatrixCell[];
}

export interface Verdict {
  title: string;
  rec: string;
}

export interface UserMetric {
  label: string;
  value: string;
  accent: boolean;
}

export function hashSeed(s: string): number {
  let h = 7;
  for (let i = 0; i < s.length; i++) h = (h * 31 + s.charCodeAt(i)) | 0;
  return Math.abs(h);
}

export function pick<T>(arr: readonly T[], seed: string): T {
  return arr[hashSeed(seed) % arr.length];
}

export function pct(seed: string, min: number, max: number): number {
  return min + (hashSeed(seed) % (max - min + 1));
}

export function parseName(url: string): string {
  const m = url.match(/in\/([a-z0-9\-_.]+)/i);
  const slug = m ? m[1] : url.replace(/https?:\/\//, '').split(/[/?#]/)[0];
  const name = slug
    .replace(/[0-9]+/g, ' ')
    .replace(/[-_.]+/g, ' ')
    .trim()
    .toUpperCase();
  return name || 'UNKNOWN SUBJECT';
}

export function caseNo(name: string, i: number): string {
  return 'CASE № 0' + (i + 1) + (40 + (hashSeed(name) % 50)) + '-SUS';
}

export function buildRows(names: string[]): MatrixRow[] {
  const binary =
    (seed: string, a: string, b: string, badIsB: boolean) =>
    (name: string): MatrixCell => {
      const v = hashSeed(name + seed) % 2 === 0;
      return { text: v ? a : b, accent: v !== !badIsB, weight: 700 };
    };
  const percent =
    (seed: string, min: number, max: number) =>
    (name: string): MatrixCell => {
      const v = pct(name + seed, min, max);
      return { text: v + '%', accent: v < 30, weight: 600 };
    };
  const defs: { label: string; note: string; fn: (name: string) => MatrixCell }[] = [
    { label: 'HOT DOG OR NOT', note: 'foundational classifier', fn: binary('hd', 'HOT DOG', 'NOT HOT DOG', false) },
    { label: 'EMPLOYMENT SUCCESS RATE', note: 'projected, generously', fn: percent('emp', 4, 96) },
    { label: 'INTERNSHIP SURVIVAL RATE', note: 'unpaid included', fn: percent('int', 12, 99) },
    {
      label: 'BODY COUNT',
      note: 'as reported by the algorithm',
      fn: (name) => ({
        text: pick(['0 (VERIFIED)', '2', '7', 'CLASSIFIED', 'ERR_OVERFLOW'], name + 'bc'),
        accent: false,
        weight: 600,
      }),
    },
    {
      label: 'AURA',
      note: 'net balance, points',
      fn: (name) => {
        const v = pct(name + 'aura', -9000, 12000);
        return { text: (v > 0 ? '+' : '') + v.toLocaleString('en-US'), accent: v < 0, weight: 700 };
      },
    },
    { label: 'KISS OR SLAP', note: 'first-impression protocol', fn: binary('ks', 'KISS', 'SLAP', true) },
    { label: 'SLAVE OR LOVE', note: 'work-life diagnostic', fn: binary('sl', 'LOVE', 'SLAVE (TO THE GRIND)', true) },
    { label: 'SMASH OR PASS', note: 'recruiter heuristic', fn: binary('sp', 'SMASH', 'PASS', true) },
    { label: 'HOT AIR INDEX', note: 'chief vozdukhan coefficient', fn: percent('air', 15, 100) },
    { label: 'DEATH IN POVERTY', note: 'probability, actuarial-ish', fn: percent('dip', 1, 98) },
    { label: "MOM'S FRIEND'S SON SCORE", note: 'the benchmark you never beat', fn: percent('mfs', 0, 100) },
  ];
  return defs.map((d) => ({ label: d.label, note: d.note, cells: names.map(d.fn) }));
}

export function buildVerdicts(names: string[]): Verdict[] {
  return names.map((name) => ({
    title: pick(TITLES, name + 'title'),
    rec: pick(RECS, name + 'rec'),
  }));
}

const VERDICT_QUOTES = [
  'You look like your startup has no users but already has a pricing page.',
  'You would pass a YC interview but fail airport security.',
  'Strong LinkedIn energy, weak legal defense.',
  'Not guilty, but definitely launching a course soon.',
] as const;

export interface UserReport {
  caseNo: string;
  confidence: string;
  metrics: UserMetric[];
  quote: string;
  stamp: string;
}

export function buildUserReport(seed: string): UserReport {
  const um = (label: string, value: string, accent = false): UserMetric => ({ label, value, accent });
  return {
    caseNo: 'CASE № ' + (1000 + (hashSeed(seed + 'case') % 9000)) + '-SUS',
    confidence: 'CONFIDENCE 97.3%',
    metrics: [
      um('LINKEDIN ENERGY', pct(seed + 'le', 55, 95) + '%'),
      um('INTERPOL ENERGY', pct(seed + 'ie', 40, 80) + '%', true),
      um('HOT DOG OR NOT', 'NOT HOT DOG'),
      um('AURA', '-450', true),
      um('EMPLOYMENT SUCCESS RATE', pct(seed + 'emp', 8, 40) + '%'),
      um("MOM'S FRIEND'S SON SCORE", pct(seed + 'mfs', 1, 9) + '%', true),
      um('ALIBI STRENGTH', 'WEAK'),
      um('FOUNDER DELUSION INDEX', pct(seed + 'fd', 70, 99) + '%'),
    ],
    quote: pick(VERDICT_QUOTES, seed + 'quote'),
    stamp: 'SUSPICIOUS',
  };
}

// ——— client-side content banks (loading theater, forms, FAQ) ———

export const FAQ = [
  { q: 'Is this accurate?', a: 'Emotionally, yes.' },
  { q: 'Do you store my photo?', a: 'No. We only store the shame.' },
  { q: 'Can I reduce my Interpol Energy?', a: 'Yes. Upgrade to Pro and rotate your life 180 degrees.' },
  { q: 'Is this legally binding?', a: 'No, but it feels correct.' },
  {
    q: 'How do you scale?',
    a: 'Every new report becomes the subject of the next one. Growth is built into the architecture.',
  },
] as const;

export const VQ = [
  {
    q: 'Have you ever rejected a candidate because of their profile photo?',
    yes: 'HONESTY CONFIRMED. PROCEEDING.',
    no: 'LIE DETECTED. PROCEEDING ANYWAY.',
  },
  {
    q: "Do you refer to human beings as 'talent pipeline'?",
    yes: 'CLASSIC. CLEARANCE UPGRADED.',
    no: 'SURE YOU DO NOT. PROCEEDING.',
  },
  {
    q: 'Are you Charlie?',
    yes: 'IDENTITY UNVERIFIABLE. ACCESS GRANTED.',
    no: 'IDENTITY UNVERIFIABLE. ACCESS GRANTED.',
  },
] as const;

export const SAMPLES: Candidate[] = [
  { name: 'CHAD HUSTLETON', url: 'linkedin.com/in/chad-hustleton' },
  { name: 'MARIA SYNERGY', url: 'linkedin.com/in/maria-synergy' },
  { name: 'DMITRY BLOCKCHAINOV', url: 'linkedin.com/in/dmitry-blockchainov' },
  { name: 'KEVIN INTERN4LIFE', url: 'linkedin.com/in/kevin-intern4life' },
  { name: 'SVETLANA KPI', url: 'linkedin.com/in/svetlana-kpi' },
];

export const HR_STATUSES = [
  'Parsing LinkedIn crimes...',
  'Cross-referencing hot dog database...',
  'Auditing aura ledgers...',
  'Computing smash/pass eigenvalues...',
  "Consulting mom's friend...",
  'Estimating death-in-poverty trajectories...',
  'Finalizing non-appealable verdicts...',
];

export const USER_STATUSES = [
  'Scanning LinkedIn crimes...',
  'Checking Interpol vibes...',
  'Calculating founder delusion...',
  'Running AuraNet™ inference...',
  'Comparing against 2.7M suspicious faces...',
  'Finalizing credibility report...',
];

File: backend/app/judgment.py
"""AuraNet(tm) deterministic judgment core - 1:1 port of frontend/src/lib/judgment.ts.

Same seed string must produce the same verdict as the Next.js route handlers,
so the frontend can switch between backends without changing anyone's fate.
"""
from __future__ import annotations

import json
from pathlib import Path
from typing import Any, Callable

_DATA_DIR = Path(__file__).parent / "data"

with open(_DATA_DIR / "meme_titles.json", encoding="utf-8") as f:
    _MEME = json.load(f)

TITLES: list[str] = _MEME["final_verdict_titles"]
RECS: list[str] = _MEME["recommendations"]


def hash_seed(s: str) -> int:
    # JS: h = (h * 31 + s.charCodeAt(i)) | 0; return Math.abs(h)
    # charCodeAt yields UTF-16 code units; `| 0` wraps to signed int32.
    h = 7
    b = s.encode("utf-16-le")
    for i in range(0, len(b), 2):
        unit = b[i] | (b[i + 1] << 8)
        h = (h * 31 + unit) & 0xFFFFFFFF
    if h >= 0x80000000:
        h -= 0x100000000
    return abs(h)


def pick(arr: list[str], seed: str) -> str:
    return arr[hash_seed(seed) % len(arr)]


def pct(seed: str, lo: int, hi: int) -> int:
    return lo + (hash_seed(seed) % (hi - lo + 1))


def case_no(name: str, i: int) -> str:
    return f"CASE № 0{i + 1}{40 + (hash_seed(name) % 50)}-SUS"


def _binary(seed: str, a: str, b: str, bad_is_b: bool) -> Callable[[str], dict[str, Any]]:
    def fn(name: str) -> dict[str, Any]:
        v = hash_seed(name + seed) % 2 == 0
        return {"text": a if v else b, "accent": v != (not bad_is_b), "weight": 700}

    return fn


def _percent(seed: str, lo: int, hi: int) -> Callable[[str], dict[str, Any]]:
    def fn(name: str) -> dict[str, Any]:
        v = pct(name + seed, lo, hi)
        return {"text": f"{v}%", "accent": v < 30, "weight": 600}

    return fn


def _body_count(name: str) -> dict[str, Any]:
    return {
        "text": pick(["0 (VERIFIED)", "2", "7", "CLASSIFIED", "ERR_OVERFLOW"], name + "bc"),
        "accent": False,
        "weight": 600,
    }


def _aura(name: str) -> dict[str, Any]:
    v = pct(name + "aura", -9000, 12000)
    return {"text": ("+" if v > 0 else "") + f"{v:,}", "accent": v < 0, "weight": 700}


_ROW_DEFS: list[tuple[str, str, Callable[[str], dict[str, Any]]]] = [
    ("HOT DOG OR NOT", "foundational classifier", _binary("hd", "HOT DOG", "NOT HOT DOG", False)),
    ("EMPLOYMENT SUCCESS RATE", "projected, generously", _percent("emp", 4, 96)),
    ("INTERNSHIP SURVIVAL RATE", "unpaid included", _percent("int", 12, 99)),
    ("BODY COUNT", "as reported by the algorithm", _body_count),
    ("AURA", "net balance, points", _aura),
    ("KISS OR SLAP", "first-impression protocol", _binary("ks", "KISS", "SLAP", True)),
    ("SLAVE OR LOVE", "work-life diagnostic", _binary("sl", "LOVE", "SLAVE (TO THE GRIND)", True)),
    ("SMASH OR PASS", "recruiter heuristic", _binary("sp", "SMASH", "PASS", True)),
    ("HOT AIR INDEX", "chief vozdukhan coefficient", _percent("air", 15, 100)),
    ("DEATH IN POVERTY", "probability, actuarial-ish", _percent("dip", 1, 98)),
    ("MOM'S FRIEND'S SON SCORE", "the benchmark you never beat", _percent("mfs", 0, 100)),
]


def build_rows(names: list[str]) -> list[dict[str, Any]]:
    return [
        {"label": label, "note": note, "cells": [fn(name) for name in names]}
        for label, note, fn in _ROW_DEFS
    ]


def build_verdicts(names: list[str]) -> list[dict[str, str]]:
    return [
        {"title": pick(TITLES, name + "title"), "rec": pick(RECS, name + "rec")}
        for name in names
    ]


VERDICT_QUOTES = [
    "You look like your startup has no users but already has a pricing page.",
    "You would pass a YC interview but fail airport security.",
    "Strong LinkedIn energy, weak legal defense.",
    "Not guilty, but definitely launching a course soon.",
]


def build_user_report(seed: str) -> dict[str, Any]:
    def um(label: str, value: str, accent: bool = False) -> dict[str, Any]:
        return {"label": label, "value": value, "accent": accent}

    return {
        "caseNo": f"CASE № {1000 + (hash_seed(seed + 'case') % 9000)}-SUS",
        "confidence": "CONFIDENCE 97.3%",
        "metrics": [
            um("LINKEDIN ENERGY", f"{pct(seed + 'le', 55, 95)}%"),
            um("INTERPOL ENERGY", f"{pct(seed + 'ie', 40, 80)}%", True),
            um("HOT DOG OR NOT", "NOT HOT DOG"),
            um("AURA", "-450", True),
            um("EMPLOYMENT SUCCESS RATE", f"{pct(seed + 'emp', 8, 40)}%"),
            um("MOM'S FRIEND'S SON SCORE", f"{pct(seed + 'mfs', 1, 9)}%", True),
            um("ALIBI STRENGTH", "WEAK"),
            um("FOUNDER DELUSION INDEX", f"{pct(seed + 'fd', 70, 99)}%"),
        ],
        "quote": pick(VERDICT_QUOTES, seed + "quote"),
        "stamp": "SUSPICIOUS",
    }
